### PR TITLE
Nomad systemd unit - Add restarting on failure

### PR DESCRIPTION
**Pull request by msvbhat**
Signed-off-by: M S Vishwanath Bhat <msvbhat@gmail.com>

### CODE REVIEW

**Comment by stevenscg** [comment]
I also have similar lines in my systemd file and it makes sense that systemd should restart the process upon failure.

File: dist/systemd/nomad.service [hunk 1]
```diff
@@ -1,12 +1,16 @@
...
+Wants=network-online.target
+After=network-online.target
 
 [Service]
 KillMode=process
 KillSignal=SIGINT
 ExecStart=/usr/bin/nomad agent -config /etc/nomad
 ExecReload=/bin/kill -HUP $MAINPID
+Restart=on-failure
+RestartSec=30
```
Comment (schmichael) [inline dist/systemd/nomad.service]: This seems awfully long. The Nomad agent must heartbeat script health checks otherwise they'll be marked as failing. Perhaps shorten this to a few seconds to prevent script checks from expiring?

**Comment by schmichael** [comment]
After discussing it with coworkers I'd suggest adding something like:

```ini
StartLimitBurst=3
StartLimitIntervalSec=10
```
https://www.freedesktop.org/software/systemd/man/systemd.unit.html#StartLimitIntervalSec=interval

[`RestartSec`](https://www.freedesktop.org/software/systemd/man/systemd.service.html#RestartSec=) appears to default to 100ms which seems ok. Let me know if there's some benefit to a longer restart delay. These settings have surprising consequences sometimes!

File: dist/systemd/nomad.service (hunk 1 above)
Comment (msvbhat) [reply dist/systemd/nomad.service]: Okay. I will remove the `RestartSec` so that it defaults to 100ms.

**Comment by msvbhat** [comment]
Thanks @schmichael. I will send updated patch with review comment changes.

**Comment by msvbhat** [comment]
Looks like some intermittent and sporadic failure. Is there a way to fix this from this patch?

File: dist/systemd/nomad.service (hunk 1 above)
Comment (shantanugadgil) [reply dist/systemd/nomad.service]: In my opinion 100 ms is way too quick. I personally use a RestartSec of 10 seconds

Comment (msvbhat) [reply dist/systemd/nomad.service]: I think default value of 100ms is Okay. May be it is on the aggressive side. But considering that this is the sleeping time before restarting the unit, the default value should be good enough.

Comment (shantanugadgil) [reply dist/systemd/nomad.service]: The reason I added the large delay was due to the scenario where sometimes the data directory gets corrupted after a machine boot.
The delay is also to let Consul startup.

Comment (schmichael) [reply dist/systemd/nomad.service]: Hm perhaps a `RestartSec=2` would be sufficient for that and mentioning how to setup the Consul dependency in a comment like:

```ini
Wants=network-online.target
After=network-online.target

# If you're using Consul you should use the following Wants/After settings instead:
#Wants=consul.service
#After=consul.service
```

Comment (msvbhat) [reply dist/systemd/nomad.service]: Okay, Will update the PR with comments to have a dependency on Consul.

Should I also have 2 seconds of delay between restarts?

Comment (schmichael) [reply dist/systemd/nomad.service]: Yeah, sounds good to me. Thanks for your patience! Any further tweaks we should open a new issue for discussion as what's in this PR is already a big improvement.

**Comment by msvbhat** [comment]
@schmichael Thanks. I have fixed the changes requested. Can you please review again?

**Comment by schmichael** [comment]
Approved and merged! Thanks for your contribution and patience @msvbhat!

Thanks to everyone else for your comments as well. It also kicked off a conversation internally at HashiCorp about standardizing all of our systemd service units across projects, so stay tuned for an even better out of the box experience in the future!

**Comment by shantanugadgil** [comment]
thanks @schmichael  it always helps to have a HashiCorp "blessed" version of something to use as reference, rather than cobbling things together from various sources. 😁

**Comment by msvbhat** [comment]
@schmichael Thanks for providing valuable comments and being kind to the first time contributor 💯
